colmem: improve memory-limiting behavior of the accounting helpers

yuzefovich opened the pull request:
**colmem: introduce a helper method when no memory limit should be applied**

This commit is a pure mechanical change.

Release note: None

**colmem: move some logic of capacity-limiting into the accounting helper**

This commit moves the logic that was duplicated across each user of the
SetAccountingHelper into the helper itself. Clearly, this allows us to
de-duplicate some code, but it'll make it easier to refactor the code
which is done in the following commit.

Additionally, this commit makes a tiny change to make the resetting
behavior in the hash aggregator more precise.

Release note: None

**colmem: improve memory-limiting behavior of the accounting helpers**

This commit fixes an oversight in how we are allocating batches of the
"dynamic" capacity. We have two related ways for reallocating batches,
and both of them work by growing the capacity of the batch until the
memory limit is exceeded, and then the batch would be reused until the
end of the query execution. This is a reasonable heuristic under the
assumption that all tuples in the data stream are roughly equal in size,
but this might not be the case.

In particular, consider an example when 10k small rows of 1KiB are
followed by 10k large rows of 1MiB. According to our heuristic, we
happily grow the batch until 1024 in capacity, and then we do not shrink
the capacity of that batch, so once the large rows start appearing, we
put 1GiB worth of data into a single batch, significantly exceeding our
memory limit (usually 64MiB with the default `workmem` setting).

This commit introduces a new heuristic as follows:
- the first time a batch exceeds the memory limit, its capacity is memorized,
  and from now on that capacity will determine the upper bound on the
  capacities of the batches allocated through the helper;
- if at any point in time a batch exceeds the memory limit by at least a
  factor of two, then that batch is discarded, and the capacity will never
  exceed half of the capacity of the discarded batch;
- if the memory limit is not reached, then the behavior of the dynamic growth
  of the capacity provided by `Allocator.ResetMaybeReallocate` is still
  applicable (i.e. the capacities will grow exponentially until
  coldata.BatchSize()).

Note that this heuristic does not have an ability to grow the maximum
capacity once it's been set although it might make sense to do so (say,
if after shrinking the capacity, the next five times we see that the
batch is using less than half of the memory limit). This is an conscious
omission since I want this change to be backported, and never growing
seems like a safer choice. Thus, this improvement is left as a TODO.

Also, we still might create batches that are too large in memory
footprint in those places that don't use the SetAccountingHelper (e.g.
in the columnarizer) since we perform the memory limit check at the
batch granularity. However, this commit improves things there so that we
don't reuse that batch on the next iteration and will use half of the
capacity on the next iteration.

Fixes: #76464.

Release note (bug fix): CockroachDB now more precisely respects the
`distsql_workmem` setting which improves the stability of each node and
makes OOMs less likely.

**colmem: unexport Allocator.ResetMaybeReallocate**

This commit is a mechanical change to unexport
`Allocator.ResetMaybeReallocate` so that the users would be forced to use
the method with the same name from the helpers. This required splitting
off the tests into two files.

Release note: None

yuzefovich committed:
colmem: introduce a helper method when no memory limit should be applied

This commit is a pure mechanical change.

Release note: None

yuzefovich committed:
colmem: move some logic of capacity-limiting into the accounting helper

This commit moves the logic that was duplicated across each user of the
SetAccountingHelper into the helper itself. Clearly, this allows us to
de-duplicate some code, but it'll make it easier to refactor the code
which is done in the following commit.

Additionally, this commit makes a tiny change to make the resetting
behavior in the hash aggregator more precise.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/85440)
<!-- Reviewable:end -->

yuzefovich committed:
colmem: improve memory-limiting behavior of the accounting helpers

This commit fixes an oversight in how we are allocating batches of the
"dynamic" capacity. We have two related ways for reallocating batches,
and both of them work by growing the capacity of the batch until the
memory limit is exceeded, and then the batch would be reused until the
end of the query execution. This is a reasonable heuristic under the
assumption that all tuples in the data stream are roughly equal in size,
but this might not be the case.

In particular, consider an example when 10k small rows of 1KiB are
followed by 10k large rows of 1MiB. According to our heuristic, we
happily grow the batch until 1024 in capacity, and then we do not shrink
the capacity of that batch, so once the large rows start appearing, we
put 1GiB worth of data into a single batch, significantly exceeding our
memory limit (usually 64MiB with the default `workmem` setting).

This commit introduces a new heuristic as follows:
- the first time a batch exceeds the memory limit, its capacity is memorized,
  and from now on that capacity will determine the upper bound on the
  capacities of the batches allocated through the helper;
- if at any point in time a batch exceeds the memory limit by at least a
  factor of two, then that batch is discarded, and the capacity will never
  exceed half of the capacity of the discarded batch;
- if the memory limit is not reached, then the behavior of the dynamic growth
  of the capacity provided by `Allocator.ResetMaybeReallocate` is still
  applicable (i.e. the capacities will grow exponentially until
  coldata.BatchSize()).

Note that this heuristic does not have an ability to grow the maximum
capacity once it's been set although it might make sense to do so (say,
if after shrinking the capacity, the next five times we see that the
batch is using less than half of the memory limit). This is an conscious
omission since I want this change to be backported, and never growing
seems like a safer choice. Thus, this improvement is left as a TODO.

Also, we still might create batches that are too large in memory
footprint in those places that don't use the SetAccountingHelper (e.g.
in the columnarizer) since we perform the memory limit check at the
batch granularity. However, this commit improves things there so that we
don't reuse that batch on the next iteration and will use half of the
capacity on the next iteration.

Release note (bug fix): CockroachDB now more precisely respects the
`distsql_workmem` setting which improves the stability of each node and
makes OOMs less likely.

yuzefovich committed:
colmem: unexport Allocator.ResetMaybeReallocate

This commit is a mechanical change to unexport
`Allocator.ResetMaybeReallocate` so that the users would be forced to use
the method with the same name from the helpers. This required splitting
off the tests into two files.

Release note: None

DrewKimball reviewed (commented):
Nice refactoring/stability improvements!

> Reviewed 7 of 7 files at r1, 8 of 8 files at r2, 27 of 27 files at r3, 4 of 4 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85440#-:-N8fNBRrDfYquqZF2bUh:b-ttz6ao)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @michae2 and @yuzefovich)

___
*[`pkg/sql/colfetcher/cfetcher.go` line 852 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85440#-N8eyQ6P4pKhEX3nrtEN:-N8eyQ6P4pKhEX3nrtEO:b-ahtzw2) ([raw file](https://github.com/cockroachdb/cockroach/blob/85f30679a052113d102ad2bfd7391bb874b4917a/pkg/sql/colfetcher/cfetcher.go#L852)):*
> ```Go
> 
> 			if cf.machine.limitHint > 0 && cf.machine.rowIdx >= cf.machine.limitHint {
> 				// If we made it to our limit hint, so output our batch early to
> ```

[nit] "If we" -> "We"
___
*[`pkg/sql/colfetcher/vectorized_batch_size_test.go` line 153 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/85440#-N8f8ZwQB5CaRJvOupSv:-N8f8ZwQB5CaRJvOupSw:be8vsce) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f2d0ff726e4aec77ca3dd444591d80797b9ea83/pkg/sql/colfetcher/vectorized_batch_size_test.go#L153)):*
> ```Go
> 	for _, tc := range []struct {
> 		// numRows and rowSize must be of the same length, with each index
> 		// specifying the number of rows of the corresponding size to be
> ```

[nit] add that the row size is in bytes.
___
*[`pkg/sql/colmem/allocator.go` line 567 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/85440#-N8fGXx5-v1UqAmsF-XX:-N8fGXx5-v1UqAmsF-XY:bo902y2) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f2d0ff726e4aec77ca3dd444591d80797b9ea83/pkg/sql/colmem/allocator.go#L567)):*
> ```Go
> 
> // Init initializes the helper.
> func (h *AccountingHelper) Init(allocator *Allocator, memoryLimit int64) {
> ```

[nit] Maybe mention here that the allocator can be shared between different components.
___
*[`pkg/sql/colmem/allocator.go` line 573 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/85440#-N8fC3fj7S_6B94bdLhn:-N8fC3fj7S_6B94bdLho:b8umtw1) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f2d0ff726e4aec77ca3dd444591d80797b9ea83/pkg/sql/colmem/allocator.go#L573)):*
> ```Go
> 		// disk spilling" scenario, but the helper should ignore that, so we
> 		// override it to the default value of the distsql_workmem variable.
> 		memoryLimit = 64 << 20 /* 64 MiB */
> ```

Would it be better to use `GetWorkMemLimit` here? I'm wondering if there could be a case where a user sets the working memory limit to `1` for whatever reason, then gets unexpected results when we revert to the default.
___
*[`pkg/sql/colmem/allocator.go` line 584 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/85440#-N8fDWxN3MgGNGFoz2MW:-N8fDWxN3MgGNGFoz2MX:bu0df6y) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f2d0ff726e4aec77ca3dd444591d80797b9ea83/pkg/sql/colmem/allocator.go#L584)):*
> ```Go
> // yet to be set, use 0 if unknown.
> func (h *AccountingHelper) ResetMaybeReallocate(
> 	typs []*types.T, oldBatch coldata.Batch, remainingTuples int,
> ```

[nit] `remainingTuples` confused me a bit when I saw this method being called elsewhere; maybe something like `tuplesToBeSet` or `remainingTuplesToSet` would be clearer?
___
*[`pkg/sql/colmem/allocator.go` line 591 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/85440#-N8fICa1-jL2Igr7rOWd:-N8fICa1-jL2Igr7rOWe:btqu5pb) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f2d0ff726e4aec77ca3dd444591d80797b9ea83/pkg/sql/colmem/allocator.go#L591)):*
> ```Go
> 		// with other components, thus, we cannot ask it directly for the batch
> 		// mem size, yet the allocator can provide a useful upper bound.)
> 		if batchMemSizeUpperBound := h.allocator.Used(); h.discardBatch(batchMemSizeUpperBound) {
> ```

Shouldn't we also be setting `h.maxCapacity` to the current batch capacity once we exceed the memory limit? (Rather than only setting it to half the current capacity once we reach double the limit). 
___
*[`pkg/sql/colmem/allocator.go` line 719 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/85440#-N8fFIcnAm7y0fnEJXPe:-N8fFIcnAm7y0fnEJXPf:b-c7xbhc) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f2d0ff726e4aec77ca3dd444591d80797b9ea83/pkg/sql/colmem/allocator.go#L719)):*
> ```Go
> // yet to be set, use 0 if unknown.
> func (h *SetAccountingHelper) ResetMaybeReallocate(
> 	typs []*types.T, oldBatch coldata.Batch, remainingTuples int,
> ```

[nit] same as for `AccountingHelper`.
___
*[`pkg/sql/colmem/allocator.go` line 821 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/85440#-N8fJ4-T7RIXIm9GivsN:-N8fJ4-T7RIXIm9GivsO:b-zao8n8) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f2d0ff726e4aec77ca3dd444591d80797b9ea83/pkg/sql/colmem/allocator.go#L821)):*
> ```Go
> 		// we update the memorized capacity. If it's the latter, then on the
> 		// following call to ResetMaybeReallocate, the batch will be discarded.
> 		h.helper.maxCapacity = rowIdx + 1
> ```

Should we be doing this in the `AccountingHelper` instead?
___
*[`pkg/sql/colmem/allocator.go` line 713 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/85440#-N8fMR7bBCP0UYa-GFHy:-N8fMR7bBCP0UYa-GFHz:bo2awrj) ([raw file](https://github.com/cockroachdb/cockroach/blob/5e964486c96f106d6da1d7facba810b22c1e6474/pkg/sql/colmem/allocator.go#L713)):*
> ```Go
> 
> // ResetMaybeReallocate is a light wrapper on top of
> // Allocator.resetMaybeReallocate (and thus has the same contract) with an
> ```

[nit] This no longer has exactly the same contract.
___
*[`pkg/sql/colmem/allocator_test.go` line 412 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/85440#-N8fLHfe1-nFYoT47Z-r:-N8fLHfe1-nFYoT47Z-s:becz7zj) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f2d0ff726e4aec77ca3dd444591d80797b9ea83/pkg/sql/colmem/allocator_test.go#L412)):*
> ```Go
> 				// Now the limit has been exceeded by too much again - a new
> 				// batch of smaller capacity must be allocated.
> 				{1, true, 100},
> ```

Could you add one last iteration like `{1, false, 1300}` to ensure we never attempt to shrink the capacity to zero?


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85440)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball and @michae2)

___
*[`pkg/sql/colfetcher/cfetcher.go` line 852 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85440#-N8eyQ6P4pKhEX3nrtEN:-N8fbzlJ5H7Fqm1s1ojg:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/85f30679a052113d102ad2bfd7391bb874b4917a/pkg/sql/colfetcher/cfetcher.go#L852)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

[nit] "If we" -> "We"
</blockquote></details>

Done.
___
*[`pkg/sql/colfetcher/vectorized_batch_size_test.go` line 153 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/85440#-N8f8ZwQB5CaRJvOupSv:-N8fc8wD1vgZu11FaBhB:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f2d0ff726e4aec77ca3dd444591d80797b9ea83/pkg/sql/colfetcher/vectorized_batch_size_test.go#L153)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

[nit] add that the row size is in bytes.
</blockquote></details>

Done.
___
*[`pkg/sql/colmem/allocator.go` line 567 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/85440#-N8fGXx5-v1UqAmsF-XX:-N8fcGRR3KPATE5HR9N6:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f2d0ff726e4aec77ca3dd444591d80797b9ea83/pkg/sql/colmem/allocator.go#L567)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

[nit] Maybe mention here that the allocator can be shared between different components.
</blockquote></details>

Done.
___
*[`pkg/sql/colmem/allocator.go` line 573 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/85440#-N8fC3fj7S_6B94bdLhn:-N8fcHAD9DXX_0FbRJz3:bn74uln) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f2d0ff726e4aec77ca3dd444591d80797b9ea83/pkg/sql/colmem/allocator.go#L573)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Would it be better to use `GetWorkMemLimit` here? I'm wondering if there could be a case where a user sets the working memory limit to `1` for whatever reason, then gets unexpected results when we revert to the default.
</blockquote></details>

We have a similar check in several places. This is a good point though - we should just prohibit the users from setting exactly 1B and reserve that value for internal test use. I'll open up a separate PR for that.

Another note is that I chose to hard-code the number 64MiB instead of using `execinfra.DefaultMemoryLimit` in order to not pollute the dependency graph.
___
*[`pkg/sql/colmem/allocator.go` line 584 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/85440#-N8fDWxN3MgGNGFoz2MW:-N8fdr-k7G3Bhn8ohaeZ:b6g4yxq) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f2d0ff726e4aec77ca3dd444591d80797b9ea83/pkg/sql/colmem/allocator.go#L584)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

[nit] `remainingTuples` confused me a bit when I saw this method being called elsewhere; maybe something like `tuplesToBeSet` or `remainingTuplesToSet` would be clearer?
</blockquote></details>

I like `tuplesToBeSet`, done.
___
*[`pkg/sql/colmem/allocator.go` line 591 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/85440#-N8fICa1-jL2Igr7rOWd:-N8fh4KL5OG4BrjQsEYZ:bujqsk) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f2d0ff726e4aec77ca3dd444591d80797b9ea83/pkg/sql/colmem/allocator.go#L591)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Shouldn't we also be setting `h.maxCapacity` to the current batch capacity once we exceed the memory limit? (Rather than only setting it to half the current capacity once we reach double the limit). 
</blockquote></details>

That's a good point. This is not strictly necessary for the correct behavior because `Allocator.ResetMaybeReallocate` would never allocate a new batch once it realizes that the old batch has reached the memory limit, but explicitly memorizing the max capacity in such a scenario makes things more clear (as well as avoid computing the batch memory size). Done.
___
*[`pkg/sql/colmem/allocator.go` line 719 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/85440#-N8fFIcnAm7y0fnEJXPe:-N8fhPnyAc9Otb4nbLN3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f2d0ff726e4aec77ca3dd444591d80797b9ea83/pkg/sql/colmem/allocator.go#L719)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

[nit] same as for `AccountingHelper`.
</blockquote></details>

Done.
___
*[`pkg/sql/colmem/allocator.go` line 821 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/85440#-N8fJ4-T7RIXIm9GivsN:-N8fiSRy7uU14zUT421o:bl2yo4s) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f2d0ff726e4aec77ca3dd444591d80797b9ea83/pkg/sql/colmem/allocator.go#L821)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Should we be doing this in the `AccountingHelper` instead?
</blockquote></details>

I think in `SetAccountingHelper` we can be more precise since `AccountingHelper` operates at a batch granularity. Consider an example when the batch has capacity of 16, but the `SetAccountingHelper` realizes that the memory limit is reached only after 3 tuples are set. The `AccountingHelper` could memorize 16 as `maxCapacity` which would not work well with the expectations of the `SetAccountingHelper` - its user should now be setting at most 3 tuples.

The way I think about it is that the `AccountingHelper` works at a batch granularity, and `SetAccountingHelper` is a specialization of the `AccountingHelper` where we work at a tuple granularity, so the latter can make better decisions.

Does this make sense?
___
*[`pkg/sql/colmem/allocator.go` line 713 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/85440#-N8fMR7bBCP0UYa-GFHy:-N8fiPYZ8BuUEjr2htK5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5e964486c96f106d6da1d7facba810b22c1e6474/pkg/sql/colmem/allocator.go#L713)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

[nit] This no longer has exactly the same contract.
</blockquote></details>

Done.
___
*[`pkg/sql/colmem/allocator_test.go` line 412 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/85440#-N8fLHfe1-nFYoT47Z-r:-N8fjq86C9L_i_oyz4Ou:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f2d0ff726e4aec77ca3dd444591d80797b9ea83/pkg/sql/colmem/allocator_test.go#L412)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Could you add one last iteration like `{1, false, 1300}` to ensure we never attempt to shrink the capacity to zero?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85440#-:-N8fom7k5gS237RqNymb:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @DrewKimball, @michae2, and @yuzefovich)

___
*[`pkg/sql/colmem/allocator.go` line 821 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/85440#-N8fJ4-T7RIXIm9GivsN:-N8fni7P-mdhXTFFA5Ct:b-ukq957) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f2d0ff726e4aec77ca3dd444591d80797b9ea83/pkg/sql/colmem/allocator.go#L821)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

I think in `SetAccountingHelper` we can be more precise since `AccountingHelper` operates at a batch granularity. Consider an example when the batch has capacity of 16, but the `SetAccountingHelper` realizes that the memory limit is reached only after 3 tuples are set. The `AccountingHelper` could memorize 16 as `maxCapacity` which would not work well with the expectations of the `SetAccountingHelper` - its user should now be setting at most 3 tuples.

The way I think about it is that the `AccountingHelper` works at a batch granularity, and `SetAccountingHelper` is a specialization of the `AccountingHelper` where we work at a tuple granularity, so the latter can make better decisions.

Does this make sense?
</blockquote></details>

Yeah, SGTM.


<!-- Sent from Reviewable.io -->

yuzefovich commented:
I ran [microbenchmarks](https://gist.github.com/yuzefovich/b5a6e68c45e42a79345ed5962830676b) of `colexec` package, and they mostly show minor speedup. The only concerningly looking result is for `OrderedSynchronizer`, but when running it in isolation, I see no difference, so I think that outlier can be ignored:
```
Switching to HEAD^^^^
+ make bench PKG=./pkg/sql/colexec BENCHTIMEOUT=24h BENCHES=BenchmarkOrderedSynchronizer 'TESTFLAGS=-count 10 -benchmem'
Switching to HEAD
+ make bench PKG=./pkg/sql/colexec BENCHTIMEOUT=24h BENCHES=BenchmarkOrderedSynchronizer 'TESTFLAGS=-count 10 -benchmem'
name                    old time/op    new time/op    delta
OrderedSynchronizer-24    64.6µs ± 8%    61.9µs ± 6%   ~     (p=0.089 n=10+10)

name                    old speed      new speed      delta
OrderedSynchronizer-24   381MB/s ± 8%   398MB/s ± 6%   ~     (p=0.089 n=10+10)

name                    old alloc/op   new alloc/op   delta
OrderedSynchronizer-24     1.00B ± 0%     1.00B ± 0%   ~     (all equal)

name                    old allocs/op  new allocs/op  delta
OrderedSynchronizer-24      0.00           0.00        ~     (all equal)
```

This PR will benefit from an extra set of eyes since I want to backport this to 22.1, so I'll wait for @michae2 to take a look too.

yuzefovich commented:
@michae2 will you have time to take a look at this this week - that's ok if not? I'd like to merge it before stability, and I'm off on Th-F.

michae2 reviewed (commented):
Nice work! I'm still reading.

> This commit introduces a new heuristic as follows:
> 
> the first time a batch exceeds the memory limit, its capacity is memorized,
> and from now on that capacity will determine the upper bound on the
> capacities of the batches allocated through the helper;
> if at any point in time a batch exceeds the memory limit by at least a
> factor of two, then that batch is discarded, and the capacity will never
> exceed half of the capacity of the discarded batch;
> if the memory limit is not reached, then the behavior of the dynamic growth
> of the capacity provided by Allocator.ResetMaybeReallocate is still
> applicable (i.e. the capacities will grow exponentially until
> coldata.BatchSize()).
> 
> Note that this heuristic does not have an ability to grow the maximum
> capacity once it's been set although it might make sense to do so (say,
> if after shrinking the capacity, the next five times we see that the
> batch is using less than half of the memory limit). This is an conscious
> omission since I want this change to be backported, and never growing
> seems like a safer choice. Thus, this improvement is left as a TODO.

Sounds a lot like [TCP congestion control](https://en.wikipedia.org/wiki/TCP_congestion_control). Maybe once we've hit the memory limit, we could use [additive increase / multiplicative decrease](https://en.wikipedia.org/wiki/Additive_increase/multiplicative_decrease) instead of staying at that fixed capacity? But that's just an idea, this is already an improvement over what we've got. A TODO is fine. I'll keep reading.

> Reviewed 6 of 7 files at r1, 29 of 29 files at r5, 2 of 8 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85440#-:-N946i801TuF3t5ld-SM:b5uvrtq)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @DrewKimball, @michae2, and @yuzefovich)



<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> Sounds a lot like TCP congestion control. Maybe once we've hit the memory limit, we could use additive increase / multiplicative decrease instead of staying at that fixed capacity?

Indeed, I also noticed a similar pattern. However, I realized that here we cannot really apply the AIMD pattern - the cost for that "additive" increase is non trivial since we would have to reallocate the whole underlying memory rather than "adding" a little more. Consider an example when we already have the capacity for 32 rows, and if our additive is 2 rows, then we'd release the whole 32-row batch to allocate 34-row one - that reallocation has non-trivial implications, especially so if we never reach an equilibrium (because then we'd be continuously reallocating memory - we would rather stick to a smaller batch and never reallocate again (unless that batch is extremely small)).

(I guess we could use additive increase when the underlying capacities of the vectors allow that - i.e. say we asked for `make([]int64, 10)`, but got a slice of capacity 16 - we could use the additive a few times at no cost. However, that would not be trivial for varlen types likes Bytes.)

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85440#-:-N948Dyn3ZVB0dilGcDV:b-vt83k6)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @DrewKimball, @michae2, and @yuzefovich)



<!-- Sent from Reviewable.io -->

DrewKimball commented:
I feel like the 'right' way to solve this in the long term might be to make the various `Set`, `Copy` and `Append` batch and `Vec` methods deterministic in terms of their affect on the memory footprint of the batch. That way, we could check before performing the operation whether it would cause us to exceed the limit and emit immediately if necessary. I guess for `Bytes` columns that would come down to using `make` and `copy` to grow the slice instead of `append`, and other types should be easier.

yuzefovich commented:
> I feel like the 'right' way to solve this in the long term might be to make the various Set, Copy and Append methods deterministic in terms of their affect on the memory footprint of the batch. That way, we could check before performing the operation whether it would cause us to exceed the limit and emit immediately if necessary. I guess for Bytes columns that would come down to using make and copy to grow the slice instead of append, and other types should be easier.

It's an interesting idea that'd be worth exploring. We'd need to be conscious about the performance implications though - I'm afraid that introducing those checks would have non-negligible overhead. However, at the same time those conditionals are likely to get very good correct branch prediction most of the time (once the batch has grown to maximum capacity), so maybe the impact would be relatively small. Still we would probably lose the ability to inline most of the `Set` calls.

DrewKimball commented:
> Still we would probably lose the ability to inline most of the Set calls.

I think only bytes columns would require any changes in `Set`, right? Value types like int would have no change, while pointer types like decimal or datum would just add the size of the object, so should also be deterministic.

yuzefovich commented:
> I think only bytes columns would require any changes in Set, right? Value types like int would have no change, while pointer types like decimal or datum would just add the size of the object, so should also be deterministic.

Yeah, fair point.

michae2 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> AFAICT this is all correct. Intuitively I feel like there is room for simplification in `Allocator.resetMaybeReallocate` and `AccountingHelper.ResetMaybeReallocate` but it will require more thought to discover.

Can we let this bake for a little while before backporting?

> Reviewed 6 of 8 files at r6, 27 of 27 files at r7, 4 of 4 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85440#-:-N98z6sGDx05E3eWO2Yp:bniqwdl)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @DrewKimball and @yuzefovich)

___
*[`pkg/sql/colexec/columnarizer.go` line 222 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/85440#-N98pUD3BauNj7PPoYfC:-N98pUD3BauNj7PPoYfD:b-ctz03i) ([raw file](https://github.com/cockroachdb/cockroach/blob/8abc47851882ecf681438ae12c48930ab63e358e/pkg/sql/colexec/columnarizer.go#L222)):*
> ```Go
> 	if reallocated {
> 		oldRows := c.buffered
> 		newRows := make(rowenc.EncDatumRows, c.batch.Capacity())
> ```

Should this sometimes be `c.helper.maxCapacity` instead of `c.batch.Capacity()`?


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
Thanks for the reviews! I agree with the premise that we should be able to find a solution that is both simpler and more precise.

> Can we let this bake for a little while before backporting?

For sure - especially so that we'll test this behavior as part of QA next week.

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85440#-:-N992y5K7yxxFvheRTjU:b-6yk8co)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @DrewKimball, @michae2, and @yuzefovich)

___
*[`pkg/sql/colexec/columnarizer.go` line 222 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/85440#-N98pUD3BauNj7PPoYfC:-N992Zjv5ghHoig4Leay:bca01qv) ([raw file](https://github.com/cockroachdb/cockroach/blob/8abc47851882ecf681438ae12c48930ab63e358e/pkg/sql/colexec/columnarizer.go#L222)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Should this sometimes be `c.helper.maxCapacity` instead of `c.batch.Capacity()`?
</blockquote></details>

No, I don't think so. In the columnarizer we use the `AccountingHelper` which operates at a batch-granularity, so we won't ever have a case when `c.helper.maxCapacity < c.batch.Capacity()` - `ResetMaybeReallocate` will always give us such a batch that we can use the full capacity. Only when using the `SetAccountingHelper` can we get the behavior where the helper tells us that the batch is done before its full capacity is used.


<!-- Sent from Reviewable.io -->

yuzefovich commented:
bors r-

I think this needs a rebase.

yuzefovich commented:
bors r+

michae2 reviewed (commented):
> However, I realized that here we cannot really apply the AIMD pattern - the cost for that "additive" increase is non trivial since we would have to reallocate the whole underlying memory rather than "adding" a little more. Consider an example when we already have the capacity for 32 rows, and if our additive is 2 rows, then we'd release the whole 32-row batch to allocate 34-row one - that reallocation has non-trivial implications, especially so if we never reach an equilibrium (because then we'd be continuously reallocating memory - we would rather stick to a smaller batch and never reallocate again (unless that batch is extremely small)).

Oh, good point, heh! :slightly_smiling_face:

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85440#-:-N9A0ndICmTuR9kDuS3L:b-gth537)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @DrewKimball, @michae2, and @yuzefovich)

___
*[`pkg/sql/colexec/columnarizer.go` line 222 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/85440#-N98pUD3BauNj7PPoYfC:-N9A0T6YCF5DsIgWtIzs:b-guwrpx) ([raw file](https://github.com/cockroachdb/cockroach/blob/8abc47851882ecf681438ae12c48930ab63e358e/pkg/sql/colexec/columnarizer.go#L222)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

No, I don't think so. In the columnarizer we use the `AccountingHelper` which operates at a batch-granularity, so we won't ever have a case when `c.helper.maxCapacity < c.batch.Capacity()` - `ResetMaybeReallocate` will always give us such a batch that we can use the full capacity. Only when using the `SetAccountingHelper` can we get the behavior where the helper tells us that the batch is done before its full capacity is used.
</blockquote></details>

Ah, thank you. This clarifies things for me.


<!-- Sent from Reviewable.io -->
